kbs/src/resource/pkcs11.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright (c) 2024 by IBM.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{bail, Result};
+use cryptoki::context::{CInitializeArgs, Pkcs11};
+use cryptoki::object::{Attribute, AttributeInfo, AttributeType, KeyType, ObjectClass};
+use cryptoki::session::{Session, UserType};
+use cryptoki::types::AuthPin;
+use serde::Deserialize;
+use std::sync::Arc;
+use tokio::sync::Mutex;
+
+use super::{Repository, ResourceDesc};
+
+#[derive(Debug, Deserialize, Clone)]
+pub struct Pkcs11Config {
+    /// Path to the Pkcs11 module
+    module: String,
+
+    /// The index of the slot to be used
+    /// If not provided, the first slot will be used.
+    slot_index: Option<u8>,
+
+    /// The user pin for authenticating the session
+    pin: String,
+}
+
+pub struct Pkcs11Backend {
+    session: Arc<Mutex<Session>>,
```

Comment:
Do we need `Mutex` here to control race condition? I noticed that all methods of `Session` in `cryptoki` are declared `&self` rather than `&mut self`

Reply:
I think the session struct itself is not thread safe. Without this I get 

    within `cryptoki::session::Session`, the trait `Sync` is not implemented for `*mut u32`, which is required by `Pkcs11Backend: Sync`
    
I'm not sure that this is the best way to resolve it. I don't know exactly what the properties of that field are so I went wiith a mutex.

Reply:
Makes sense to me. Based on this, I suggest `RwLock` to replace `Mutex` while `&self` operations will not be blocked.